python/message.c

```diff
@@ -0,0 +1,1546 @@
+    return false;
+  }
+  const upb_msgdef* m1_msgdef = _PyUpb_CMessage_GetMsgdef(m1);
+#ifndef NDEBUG
+  const upb_msgdef* m2_msgdef = _PyUpb_CMessage_GetMsgdef(m2);
+  assert(m1_msgdef == m2_msgdef);
+#endif
+  const upb_msg* m1_msg = PyUpb_CMessage_GetIfWritable((PyObject*)m1);
+  const upb_msg* m2_msg = PyUpb_CMessage_GetIfWritable(_m2);
+  return PyUpb_Message_IsEqual(m1_msg, m2_msg, m1_msgdef);
+}
+
+/*
+ * PyUpb_CMessage_AssureWritable()
+ *
+ * This implements the "expando" behavior of Python protos:
+ *   foo = FooProto()
+ *
+ *   # The intermediate messages don't really exist, and won't be serialized.
+ *   x = foo.bar.bar.bar.bar.bar.baz
+ *
+ *   # Now all the intermediate objects are created.
+ *   foo.bar.bar.bar.bar.bar.baz = 5
+ *
+ * This function should be called before performing any mutation of a protobuf
+ * object.
+ *
+ * Post-condition:
+ *   PyUpb_CMessage_IsUnset(self) is false
+ */
+void PyUpb_CMessage_AssureWritable(PyUpb_CMessage* self) {
+  if (!PyUpb_CMessage_IsUnset(self)) return;
+
+  // This is a non-present message. We need to create a real upb_msg for this
+  // object and every parent until we reach a present message.
+  upb_arena* arena = PyUpb_Arena_Get(self->arena);
+  PyUpb_CMessage* child = self;
+  PyUpb_CMessage* parent = self->parent;
+  const upb_fielddef* child_f = PyUpb_CMessage_GetFieldDef(child);
+  // This overwrites child->parent.
+  child->msg = upb_msg_new(upb_fielddef_msgsubdef(child_f), arena);
+
+  while (PyUpb_CMessage_IsUnset(child)) {
+    PyUpb_CMessage* next_parent = parent->parent;
+    const upb_fielddef* parent_f = NULL;
+    if (PyUpb_CMessage_IsUnset(parent)) {
+      parent_f = PyUpb_CMessage_GetFieldDef(parent);
+      // This overwrites parent->parent.
+      parent->msg = upb_msg_new(upb_fielddef_msgsubdef(parent_f), arena);
+    }
+    upb_msgval msgval;
+    msgval.msg_val = child->msg;
+    upb_msg_set(parent->msg, child_f, msgval, arena);
+    child->def = (uintptr_t)upb_fielddef_msgsubdef(child_f);
+    PyUpb_WeakMap_Delete(parent->unset_subobj_map, child_f);
+    PyUpb_ObjCache_Add(child->msg, &child->ob_base);
+    if (child != self) {
+      Py_DECREF(child);  // Was previously a parent.
+    }
+    child = parent;
+    child_f = parent_f;
+    parent = next_parent;
+  }
+
+  Py_DECREF(child);
+  self->version++;
+}
+
+static void PyUpb_CMessage_SyncSubobjs(PyUpb_CMessage* self);
+
+/*
+ * PyUpb_CMessage_SwitchToSet()
+ *
+ * The message equivalent of PyUpb_*Container_SwitchToSet(), this transitions
+ * the wrapper from the unset state (owning a reference on self->parent) to the
+ * set state (having a non-owning pointer to self->msg).
+ */
+static void PyUpb_CMessage_SwitchToSet(PyUpb_CMessage* self,
+                                       const upb_fielddef* f, upb_msg* msg) {
+  assert(f == PyUpb_CMessage_GetFieldDef(self));
+  PyUpb_ObjCache_Add(msg, &self->ob_base);
+  Py_DECREF(&self->parent->ob_base);
+  self->msg = msg;  // Overwrites self->parent
```

Comment:
it took me a really long time to understand that msg and parent were in a union (hence the comments about the overwrite).  Does C support named unions?  That might be a good way to force the code to reflect this fact explicitly.